### ISSUE TITLE
Xwayland app is entirely black after unlock from idle mode

### ISSUE THREAD

**Issue by dingiso** [swaywm/sway#7881]
- **Sway Version:**  sway version 1.9-dev-bf2b79b2 (Dec 11 2023, branch 'master')
- **Debug Log:**
```
00:00:00.000 [sway/main.c:338] Sway version 1.9-dev-bf2b79b2 (Dec 11 2023, branch 'master')
00:00:00.000 [sway/main.c:339] wlroots version 0.18.0-dev
00:00:00.002 [sway/main.c:120] Linux ArchLu 6.6.8-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 21 Dec 2023 19:01:01 +0000 x86_64 GNU/Linux
00:00:00.002 [sway/main.c:136] Contents of /etc/os-release:
00:00:00.002 [sway/main.c:120] NAME="Arch Linux"
00:00:00.002 [sway/main.c:120] PRETTY_NAME="Arch Linux"
00:00:00.002 [sway/main.c:120] ID=arch
00:00:00.002 [sway/main.c:120] BUILD_ID=rolling
00:00:00.002 [sway/main.c:120] ANSI_COLOR="38;2;23;147;209"
00:00:00.002 [sway/main.c:120] HOME_URL="https://archlinux.org/"
00:00:00.002 [sway/main.c:120] DOCUMENTATION_URL="https://wiki.archlinux.org/"
00:00:00.002 [sway/main.c:120] SUPPORT_URL="https://bbs.archlinux.org/"
00:00:00.002 [sway/main.c:120] BUG_REPORT_URL="https://bugs.archlinux.org/"
00:00:00.002 [sway/main.c:120] PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
00:00:00.002 [sway/main.c:120] LOGO=archlinux-logo
00:00:00.002 [sway/main.c:108] LD_LIBRARY_PATH=
00:00:00.002 [sway/main.c:108] LD_PRELOAD=
00:00:00.002 [sway/main.c:108] PATH=/data/.local/arm-none-eabi/bin:/home/dingisoul/.local/bin:/home/dingisoul/.npm-global/bin:/home/dingisoul/.miniconda/bin:/home/dingisoul/.miniconda/condabin:/data/.local/arm-none-eabi/bin:/home/dingisoul/.local/bin:/home/dingisoul/.nvm/versions/node/v19.9.0/bin:/home/dingisoul/dev/STM32CubeProg/bin:/home/dingisoul/.cargo/bin:/home/dingisoul/.npm-global/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
00:00:00.002 [sway/main.c:108] SWAYSOCK=/run/user/1000/sway-ipc.1000.1155.sock
00:00:00.002 [sway/main.c:346] Detected both options and positional arguments. If you are trying to use the IPC client, options are not supported. Otherwise, check the provided arguments for issues. See `man 1 sway` or `sway -h` for usage. If you are trying to generate a debug log, use `sway -d 2>sway.log`
```
- **Configuration File:** [sway.config](https://gist.github.com/dingiso/cf36e07445f1bd4de78062c6c51084e5)

- **Description:**

Dear sway maintainers,

I have a issue regarding xorg-xwayland. My xwayland application appears entirely black after resuming (unlock) from idle mode. Conversely, other native Wayland applications display correctly. I'm using Sway as my window manager, Swaylock for locking my screen, and Swayidle for managing idle mode. Here are the versions of these applications:

```
xorg-xwayland 23.2.3-1
swayidle 1.8.0-1 
swaylock 1.7.2-1
```
The following codes are my configuration's lines of idle in sway config file.
```
exec swayidle -w \
         timeout 300 'swaylock -f -c 000000' \ 
         timeout 600 'swaymsg "output * dpms off"' resume 'swaymsg "output * dpms on"' \
         before-sleep 'swaylock -f -c 000000'
```
Xwayland process:
```
1276       1  0 13:39 tty1     00:00:02 Xwayland :0 -rootless -core -terminate 10 -listenfd 37 -listenfd 39 -displayfd 87 -wm 84
```

The image below illustrates what takes place
![xwayland](https://s3.bmp.ovh/imgs/2023/12/25/cd0c2dc41d6c1a49.png)

## Other

The recent occurrence of this issue, involving displaying black windows in xwayland apps after unlocking, surfaced about two weeks ago following a regular day's software update. Before it,  there is similar issue: glitch in xwayland apps after unlocking which could be "slightly" solved using full screen to refresh the window.

Initially, I suspected that the problem might stem from insufficient space in the swap file compared to the available RAM. This could potentially lead to data loss in the memory when the system is locked. To address this issue, I increased the size of the swap file, making it larger than my memory. Despite this change, the black screen in xwayland apps persists black after unlocking.

I have been troubled by this question for a long time.

Thank you for addressing my issue. If you require more context or have suggestions on resolving this or where I should seek guidance, please kindly leave a comment

Currently, I need to kill the Xwayland process and rerun all xwayland apps to solve this.

Best Regards,
Dingisoul

**Comment by dingiso** [comment]
After some studying. The thing is related to the intel gpu driver. 
I have remove the `xf86-video-intel` and use the firmware provided by intel. 
A new issue comes. The Intel integrated GPU hangs in idle mode and doesn't resume. Additionally, the external monitor shows only a backlight, with no content displayed.

The logs related to i915 driver.

Jan 02 19:03:09 ArchLu kernel: i915 0000:00:02.0: [drm] GPU HANG: ecode 12:1:85dffffb, in Xwayland [1292]
Jan 02 19:03:09 ArchLu kernel: i915 0000:00:02.0: [drm] Xwayland[1292] context reset due to GPU hang
Jan 02 19:03:18 ArchLu kernel: i915 0000:00:02.0: [drm] *ERROR* GT0: GUC: Engine reset failed on 0:0 (rcs0) because 0x00000000
Jan 02 19:03:18 ArchLu kernel: i915 0000:00:02.0: [drm] GPU HANG: ecode 12:1:84dffffb, in Xwayland [1292]
Jan 02 19:03:18 ArchLu kernel: i915 0000:00:02.0: [drm] Resetting chip for GuC failed to reset engine mask=0x1
Jan 02 19:03:18 ArchLu kernel: i915 0000:00:02.0: [drm] Renderer[1638] context reset due to GPU hang
Jan 02 19:03:18 ArchLu kernel: i915 0000:00:02.0: [drm] Xwayland[1292] context reset due to GPU hang
Jan 02 19:03:18 ArchLu kernel: i915 0000:00:02.0: [drm] GT0: GuC firmware i915/tgl_guc_70.bin version 70.13.1
Jan 02 19:03:18 ArchLu kernel: i915 0000:00:02.0: [drm] GT0: HuC firmware i915/tgl_huc.bin version 7.9.3
Jan 02 19:03:18 ArchLu kernel: i915 0000:00:02.0: [drm] GT0: HuC: authenticated for all workloads
Jan 02 19:03:18 ArchLu kernel: i915 0000:00:02.0: [drm] GT0: GUC: submission enabled
Jan 02 19:03:18 ArchLu kernel: i915 0000:00:02.0: [drm] GT0: GUC: SLPC enabled
Jan 02 19:03:29 ArchLu kernel: Asynchronous wait on fence 0000:00:02.0:sway[1179]:2d55c6 timed out (hint:intel_atomic_commit_ready [i915])
Jan 02 19:03:29 ArchLu kernel: Asynchronous wait on fence 0000:00:02.0:sway[1179]:2d55c2 timed out (hint:intel_atomic_commit_ready [i915])

There is one section in arch wiki dedicated to this issue [https://wiki.archlinux.org/title/Intel_ … tel_driver](https://wiki.archlinux.org/title/Intel_graphics#X_freeze/crash_with_intel_driver). However, it seems impact the performance, I'll consider trying it later if other solutions don't resolve the problem

I will update in this issue if there is update

**Comment by ckudera** [comment]
I have exactly the same issue. Please let me know if you need additional logs from me.

**Comment by nbeaver** [comment]
Probable upstream bug report (same GPU hang error code): https://gitlab.freedesktop.org/mesa/mesa/-/work_items/11429
